tests/media/client/ipc/ipcClient/IpcClientTest.cpp

```diff
@@ -99,38 +99,36 @@ TEST_F(IpcClientTest, UnexpectedDisconnectWithNotification)
     expectCreateChannel();
 
     // Exit the ipc loop, simulates an unexpected disconnect
-    int32_t ipcChannelCount = 0;
-    bool isProcessed = false;
+    bool connectionBrokenCallbackCalled = false;
     auto connectionObserverMock{std::make_shared<StrictMock<firebolt::rialto::client::ConnectionObserverMock>>()};
     EXPECT_CALL(*m_channelMock, process())
         .WillOnce(Invoke(
-            [this, &ipcChannelCount, &isProcessed, &connectionObserverMock]()
+            [this, &connectionObserverMock]()
             {
-                std::unique_lock<std::mutex> locker(m_eventsLock);
                 m_sut->registerConnectionObserver(connectionObserverMock);
-                ipcChannelCount = m_channelMock.use_count();
-                isProcessed = true;
-                m_eventsCond.notify_all();
                 return false;
             }));
 
-    EXPECT_CALL(*connectionObserverMock, onConnectionBroken());
+    EXPECT_CALL(*connectionObserverMock, onConnectionBroken())
+        .WillOnce(Invoke(
+            [this, &connectionBrokenCallbackCalled]()
+            {
+                std::unique_lock<std::mutex> locker(m_eventsLock);
+                connectionBrokenCallbackCalled = true;
+                m_eventsCond.notify_all();
+            }));
     EXPECT_NO_THROW(m_sut = std::make_unique<IpcClient>(m_channelFactoryMock, m_controllerFactoryMock,
                                                         m_blockingClosureFactoryMock));
 
-    // Wait for process to set the ipcChannelCount
+    // Wait for the callback
     {
         std::unique_lock<std::mutex> locker(m_eventsLock);
-        if (!isProcessed)
+        std::chrono::duration kMaximumWaitTime{std::chrono::seconds(5)};
+        if (!connectionBrokenCallbackCalled)
         {
-            m_eventsCond.wait(locker);
+            m_eventsCond.wait_for(locker, kMaximumWaitTime);
```

Comment:
Check return status of wait_for, and fail test on timeout:
EXPECT_NE(status, std::cv_status::timeout)

Reply:
Possible race condition here - connectionBrokenCallbackCalled + notify_all() call can be called between if check and wait_for call. To fix this, you can remove if (!connectionBrokenCallbackCalled) and use wait_for with lambda:
m_eventsCond.wait_for(locker, kMaximumWaitTime, [&](){ return connectionBrokenCallbackCalled; });

Reply:
Doesn't the mutex on lines 116 and 125 prevent a race?